File: Python/Algorithms/1365.py
class Solution:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        nl = len(nums) - 1
        # sort and return values enumerated
        # we use the enumeration to calculate the
        # items to the right of the sorted value in the
        # sorted list. O(nlogn)
        enumeration = enumerate(sorted(nums, reverse=True))
        counts = {v: nl - ind for ind, v in enumeration}
        # build counts O(n)
        return [counts[v] for v in nums]

File: Rust/Algorithms/1365.rs
use std::collections::HashMap;

impl Solution {
    pub fn smaller_numbers_than_current(nums: Vec<i32>) -> Vec<i32> {
        // boilerplate set-up
        let mut seen = 0;
        let mut counts = HashMap::new();
        let mut sorted = nums.to_vec();
        sorted.sort_unstable();
        
        // build counts
        for val in &sorted {
            if (counts.contains_key(val)) {
                seen += 1;
            } else {
                counts.insert(val.clone(), seen);
                seen += 1;
            }
        }
        
        // re-use vec instead of creating new one.
        sorted.clear();
        for num in &nums {
            sorted.push(*counts.get(num).expect("The unexpected."));
        }
        
        return sorted
    }
}
